fix: initialize gpio_keys input from pin state

Finwood opened the pull request:
Some kind of race condition caused the gpio_keys input module to occasionally miss an event after boot. This is fixed by initializing the pin state variable in the input_gpio_keys module.

Related to https://github.com/starcopter/bms-firmware/issues/6

Finwood commented:
This went a bit fast, I intended to create the pull request in my fork 🙈

fabiobaltieri commented:
Sure not prob, though I'm a bit confused by the issue, I know this driver blasts the initial state at init but I'm realizing now it's not an explicit decision? Is it doing it because `gpio_pin_configure_dt` activates a pin pullup and that may be weak enough to raise over the change threshold after `gpio_pin_interrupt_configure_dt` enabled the interrupt? That's somewhat annoying because then that behavior depends on the hardware, then maybe we should just either add a delay for the pin to settle before enabling the interrupt (i.e. call `gpio_pin_configure_dt` on all pins, sleep, then set the interrupt) or explicit call `k_work_reschedule` after init to make sure all pin inital state is initialized. Or maybe add the delay AND do the initial schedule if a certain Kconfig option is enabled, so one can decide if they want the initial state at boot or not.

Let me know if you want to look into that, I'll probably do it myself otherwise now that you found this issue.

Finwood commented:
Let's see. I'm _very_ new at Zephyr, so I might not fully know what I'm doing.

This fix is in response to a bug (?) I found with the gpio_keys input module. In a project I cannot share publicly I found that when you configure an input like below, the firmware _sometimes_ skips the button _press_ event after boot.

```dts
    gpio_keys: gpio_keys {
        compatible = "gpio-keys";
        btn: btn {
            label = "External Button";
            gpios = <&gpioa 0 GPIO_ACTIVE_HIGH>;
            zephyr,code = <INPUT_KEY_0>;
        };
    };
```

I've written a minimal sample that just prints button events to the console.

```c
const struct device *const gpio_keys_dev = DEVICE_DT_GET(DT_PATH(gpio_keys));

static void gpio_keys_cb_handler(struct input_event *evt) {
	char key;
	switch (evt->code) {
	case INPUT_KEY_0:
		key = '0';
		break;
	default:
		key = '?';
		break;
	}

	printk("%9u: %c%c (dev=%s, code=%u, value=%d, sync=%u, type=%u)\n", k_uptime_get_32(), key,
	       evt->value ? '+' : '-', evt->dev->name, evt->code, evt->value, evt->sync, evt->type);
}

INPUT_CALLBACK_DEFINE(gpio_keys_dev, gpio_keys_cb_handler);
```

Occasionally (it's a bit difficult to reproduce) Zephyr misses the first button press event, with only the _release_ event showing up on the console.

```
*** Booting Zephyr OS build v3.6.0 ***

     6387: 0- (dev=gpio_keys, code=11, value=0, sync=1, type=1)
     9550: 0+ (dev=gpio_keys, code=11, value=1, sync=1, type=1)
     9591: 0- (dev=gpio_keys, code=11, value=0, sync=1, type=1)
```

@MarkoSagadin actually was the one to look into this issue and propose the fix, I only added it and accidentally opened this PR.

---

@fabiobaltieri, I hope this helps. Again, I literally started using Zephyr last week, so I will be of little help digging down on this issue.

fabiobaltieri commented:
Got it, can you disclose what SoC are you using? I understand now that you are not referring to the initial boot set but I'll look into that as well, as that bothers me now that I realized it's racing :-) the two may be related.

As for your print, you should be able to get debugging information on any application by enabling `INPUT_EVENT_DUMP`, then if you have a shell you can enable it with `input dump on` or if you don't it'll just print everything all the time.

Finwood commented:
> can you disclose what SoC are you using?

Sure thing, its an STM32G491CET6.

Thanks for the debugging hint, I'll look into that 👍

fabiobaltieri commented:
Got it, ok so I dug a bit into my issue and it seems like it's not a race condition with the pullups as I suspeced, but rather something related to the interrupt firing from steady state in the i2c gpio port expander where those buttons are connected to in the specific board where I've seen the issue, my idea of adding a sleep does not fix it, but yours does, so... thanks! :-)

As for your issue (missing first trigger), I can't really see how this helps it, I'd imagine the problem being somewhere in the interrupt controller, maybe some buggy implementation of the "both edges" mode, but hey if it fixes it for you let's go with it. I don't have any G4 but I do have some F4 boards kicking around, may try to reproduce it myself if you still see the problme but I'm happy to go with this for now.

fabiobaltieri commented on drivers/input/input_gpio_keys.c:
```diff
@@ -137,7 +137,7 @@ static int gpio_keys_interrupt_configure(const struct gpio_dt_spec *gpio_spec,
 	}
 
-	cb->pin_state = -1;
+	cb->pin_state = gpio_pin_get_dt(gpio_spec);
```
Hey could you change the `gpio_ping_get` at line 65 with `gpio_pin_get_dt` as well while you are at it? Sorry I don't normally ask for unrelated changes but this is simple enough and you have a compliance error so this needs to be updated anyway :-)

Feel free to do it in the same or on a separate commit.

fabiobaltieri reviewed (commented):
Compliance check is complaining that the commit is missing the sign-off-by, can you add it? git commit --amend -s and then force push over this

MarkoSagadin commented:
> As for your issue (missing first trigger), I can't really see how this helps it, I'd imagine the problem being somewhere in the interrupt controller, maybe some buggy implementation of the "both edges" mode, but hey if it fixes it for you let's go with it.

Hey @fabiobaltieri, I think I understand why the first trigger was missed before and why the commit fixes it.

After `gpio_keys_interrupt_configure` runs, whenever a interrupt on any edge is triggered, the `gpio_keys_interrupt` executes. That function then schedules/reschedules a work item with handler set to `gpio_keys_change_deferred`.

That function is then deciding what kind of event it should report with `input_report_key` function, however that depends on the  value of the `pin_data->cb_data.pin_state`, which is set to `-1` in `gpio_keys_interrupt_configure`.

So, whenever we get inside `gpio_keys_change_deferred` for the first time (after the boot) the input event will always be emitted, since the `new_pressed != pin_data->cb_data.pin_state` will always be true.

Now, if the `gpio_keys_interrupt` after boot fires twice very quickly (more specifically, faster than the `cfg->debounce_interval_ms`, which is 30ms by default), then the first work item won't ever be processed by the workqueue thread (since the timer for it was restarted) and will thus be processed a bit later, causing the `gpio_keys_change_deferred` to emit the wrong button state.

(Correct me if I am wrong on the workqueue behavior.)


> That's somewhat annoying because then that behavior depends on the hardware, then maybe we should just either add a delay for the pin to settle before enabling the interrupt (i.e. call gpio_pin_configure_dt on all pins, sleep, then set the interrupt) or explicit call k_work_reschedule after init to make sure all pin inital state is initialized. Or maybe add the delay AND do the initial schedule if a certain Kconfig option is enabled, so one can decide if they want the initial state at boot or not.

I think that this is reasonable alternative. The current fix works since we _assume_ that when the `gpio_keys_interrupt_configure` is running the gpio pin is stable. It might as well be in the transition, so the initial `cb->pin_state` could be wrong, and we would again get the wrong event.

One implementation could also be that we init gpio calback with a handler that restarts a one-shot timer/workqueue item whenever it fires. That handler could then determine the steady state of the pin and then enable the regular deffered logic.

fabiobaltieri commented:
> Now, if the `gpio_keys_interrupt` after boot fires twice very quickly (more specifically, faster than the `cfg->debounce_interval_ms`, which is 30ms by default), then the first work item won't ever be processed by the workqueue thread (since the timer for it was restarted) and will thus be processed a bit later, causing the `gpio_keys_change_deferred` to emit the wrong button state.

That's indeed the behavior and that's how it does debouncing, it relies on the interrupt to keep firing as the signal keeps bouncing around and then the read happens after it stabilize, after the delay. The behavior you are describing should not really be a problem, it would just cause the driver to emit a code at boot, but the code will be emitted should be the stable one, and then it'll be saved in the driver state and once it changes the new one should fire correctly.

Do you see the the driver initializing with the wrong value at boot? That would explain but then I don't think the problem could be a double trigger.

> The current fix works since we _assume_ that when the `gpio_keys_interrupt_configure` is running the gpio pin is stable.

Yeah splitting the init cycle in two and adding a sleep for the pins to stabilize in addition to the current change does sound like a good idea, that said I measured the rise time on a board a I have here with the internal pullups and I get 2.6us on the port expander (it's a pcal6416) and 430ns on the on board one (it's an nrf52832), so it's probably not really necessary, but if you want to throw a 1ms sleep just in case it certainly won't hurt.

Finwood committed:
fix: initialize gpio_keys input from pin state

Some kind of race condition caused the gpio_keys input module to
occasionally miss an event after boot. This is fixed by initializing the
pin state variable in the input_gpio_keys module.

Related to https://github.com/starcopter/bms-firmware/issues/6

Signed-off-by: Lasse Fröhner <lasse@starcopter.com>
